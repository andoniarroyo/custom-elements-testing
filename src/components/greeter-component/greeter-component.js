const templateString = `
    <style>
        .greeter-text {
          font-size: 250px;
        }
    </style>
    <div>
      <span class="greeter-text"></span>
      <button id="notice-button">Notice!</button>
    </div>
`;
const template = document.createElement('template');
template.innerHTML = templateString;

export class GreeterComponent extends HTMLElement {
  static nameToGreeteAttributeName = "name-to-greete";
  static tag = "greeter-component";

  static get tag() {
    return tag;
  }

  static get observedAttributes() {
    return [GreeterComponent.nameToGreeteAttributeName];
  }

  #currentName = "...";

  constructor() {
    super();

    this.appendTemplate();
    this.updateGreeterText();
  }

  connectedCallback() {
    this.subscribeToEvents();
  }

  attributeChangedCallback(attrName, oldVal, newVal) {
    this.#currentName = this.getAttribute(GreeterComponent.nameToGreeteAttributeName);
    this.updateGreeterText();
  }

  appendTemplate() {
    let templateContent = template.content;
    document.body.appendChild(templateContent);
  }

  updateGreeterText() {
    document.querySelector(".greeter-text").innerHTML = this.#currentName;
  }

  subscribeToEvents() {
    document.querySelector("#notice-button").addEventListener('click', ()=> {
      this.dispatchEvent(
        new CustomEvent("greetingNoticed", { bubles: true, detail: { noticedName: this.#currentName} }))
    });
  }

}

customElements.define("greeter-component", GreeterComponent);