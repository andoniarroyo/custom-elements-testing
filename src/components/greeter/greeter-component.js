const templateString = `
    <style>
        .greeter-text {
          font-size: 250px;
        }
    </style>
    <div>
      <span class="greeter-text"></span>
      <button id="notice-button">Notice!</button>
    </div>
`;
const template = document.createElement('template');
template.innerHTML = templateString;

export class GreeterComponent extends HTMLElement {
  static nameToGreeteAttributeName = "name-to-greete";
  static tag = "greeter-component";

  static get observedAttributes() {
    return [GreeterComponent.nameToGreeteAttributeName];
  }

  #currentName = "...";

  connectedCallback() {
    this.appendTemplate();
    this.subscribeToEvents();
  }

  appendTemplate() {
    let templateContent = template.content.cloneNode(true);
    this.append(templateContent);
  }

  subscribeToEvents() {
    document.querySelector("#notice-button").addEventListener('click', () => {
      this.dispatchEvent(
        new CustomEvent("greetingNoticed", { bubles: true, detail: { noticedName: this.#currentName} }))
    });
  }

  attributeChangedCallback(attrName, oldVal, newVal) {
    this.#currentName = newVal;
    this.updateGreeterText(newVal);
  }

  updateGreeterText(newGreeterName) {
    document.querySelector(".greeter-text").textContent = newGreeterName;
  }

}

customElements.define("greeter-component", GreeterComponent);