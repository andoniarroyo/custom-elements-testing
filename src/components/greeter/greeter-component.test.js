import "./greeter-component";

const GREETER_COMPONENT_TAG = "greeter-component";
const GREETING_VALUE = "dummy-test";

describe("The greeter component", () => {

    beforeEach(() => {
        givenAGreeterComponent(GREETING_VALUE);
    });

    describe('on creating', () => {
        it('creates the expected DOM structure', () => {
            // binding of the name to greete (attr -> memory -> DOM)
            expect(thenTheGreeterComponent().outerHTML).toMatchSnapshot();
        });
    });

    describe('notifying events', () => {
        it('raises a greetingNoticed event on changing the name to greete', () => {
            let raisedEvent;
            givenAGreeterComponentSubscription("greetingNoticed", event => raisedEvent = event);
            whenNoticeButtonIsClicked();
            thenTheGreetingNoticedEventIsRaised(raisedEvent, GREETING_VALUE);
        });
    });

    // given
    function givenAGreeterComponent(newGreetingValue) {
        // clean up
        document.body.innerHTML = '';

        // creates the component with the new greeting value
        const element = document.createElement(GREETER_COMPONENT_TAG);
        // appends it to the DOM
        document.body.appendChild(element);

        // set up the new value for the name to greete
        element.setAttribute("name-to-greete", newGreetingValue);
    };

    function givenAGreeterComponentSubscription(eventType, callback) {
        getGreeterComponent().addEventListener(eventType, callback);
    };

    // when
    function whenNoticeButtonIsClicked() {
        document.querySelector("#notice-button").click();
    };

    // then
    function thenTheGreeterComponent() {
        return getGreeterComponent();
    };

    function thenTheGreetingNoticedEventIsRaised (event, value) {
        expect(event).toBeTruthy();
        expect(event.detail).toBeTruthy();
        expect(event.detail.noticedName).toBe(value);
    }

    // helper
    function getGreeterComponent() {
        return document.querySelector(GREETER_COMPONENT_TAG);
    };

});