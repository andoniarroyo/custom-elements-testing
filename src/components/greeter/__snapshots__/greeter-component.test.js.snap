// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The greeter component on creating creates the expected DOM structure 1`] = `
"<greeter-component name-to-greete=\\"dummy-test\\">
    <style>
        .greeter-text {
          font-size: 250px;
        }
    </style>
    <div>
      <span class=\\"greeter-text\\">dummy-test</span>
      <button id=\\"notice-button\\">Notice!</button>
    </div>
</greeter-component>"
`;
